lib/ansible/modules/network/ios/ios_ntp.py

```diff
@@ -0,0 +1,321 @@
+options:
+    server:
+        description:
+            - Network address of NTP server.
+    source_int:
+        description:
+            - Source interface for NTP packets.
+    acl:
+        description:
+            - ACL for peer/server access restricition.
+    logging:
+        description:
+            - Enable NTP logs. Data type boolean.
+        type: bool
+        default: False
+        choices: [True, False]
+    auth:
+        description:
+            - Enable NTP authentication. Data type boolean.
+        type: bool
+        default: False
+        choices: [True, False]
+    auth_key:
+        description:
+            - md5 NTP authentication key of tye 7.
+    key_id:
+        description:
+            - auth_key id. Datat type string
+    state:
+        description:
+            - Manage the state of the resource.
+        default: present
+        choices: ['present', 'absent']
+'''
+
+EXAMPLES = '''
+# Set NTP Server with parameters
+- ios_ntp:
+    server: 10.0.255.10
+    source_int: Loopback0
+    acl: NTP_ACL
+    logging: true
+    key_id: 10
+    auth_key: 15435A030726242723273C21181319000A
+    auth: true
+    state: present
+    provider: "{{ staging }}"
+'''
+
+RETURN = '''
+commands:
+    description: command sent to the device
+    returned: always
+    type: list
+    sample: ["no ntp server 10.0.255.10", "no ntp source Loopback0"]
+'''
+import re
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.network.ios.ios import get_config, load_config
+from ansible.module_utils.network.ios.ios import ios_argument_spec, check_args
+
+
+def parse_server(line, dest):
+    if dest == 'server':
+        match = re.search(r'(ntp server )(\d+\.\d+\.\d+\.\d+)', line, re.M)
+        if match:
+            server = match.group(2)
+            return server
+
+
+def parse_source_int(line, dest):
+    if dest == 'source':
+        match = re.search(r'(ntp source )(\S+)', line, re.M)
+        if match:
+            source = match.group(2)
+            return source
+
+
+def parse_acl(line, dest):
+    if dest == 'access-group':
+        match = re.search(r'(ntp access-group )(peer )?(serve )?(\S+)', line, re.M)
+        if match:
+            acl = match.group(4)
+            return acl
+
+
+def parse_logging(line, dest):
+    if dest == 'logging':
+        logging = dest
+        return logging
+
+
+def parse_auth_key(line, dest):
+    if dest == 'authentication-key':
+        match = re.search(r'(ntp authentication-key \d+ md5 )(\w+)', line, re.M)
+        if match:
+            auth_key = match.group(2)
+            return auth_key
+
+
+def parse_key_id(line, dest):
+    if dest == 'trusted-key':
+        match = re.search(r'(ntp trusted-key )(\d+)', line, re.M)
+        if match:
+            auth_key = match.group(2)
+            return auth_key
+
+
+def parse_auth(dest):
+    if dest == 'authenticate':
+            return dest
+
+
+def map_config_to_obj(module):
```

Comment:
Overall suggestion: Matching the structure for `want` and `have` would make the map_obj_to_commands method much simpler. 

From your code, the `want` dict would always have the all the keys in it, even the ones with no value(i.e., `None`). 

If you follow the same logic for `have`, you can remove all the try-except blocks in map_obj_to_commands. Though redundant, but for the sake of readability, if you wish you can return `None` from the parse_* methods if no match is found. 

Have a look at [this](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/network/ios/ios_user.py#L352-L356)

Reply:
@NilashishC I was thinking to implement the following changes:

```
def map_params_to_obj(module):
    obj = []
    obj_dict = {}

    if module.params['state']:
        obj_dict['state'] = module.params['state']
    if module.params['server']:
        obj_dict['server'] = module.params['server']
    if module.params['source_int']:
        obj_dict['source_int'] = module.params['source_int']
    if module.params['logging']:
        obj_dict['logging'] = module.params['logging']
    if module.params['acl']:
        obj_dict['acl'] = module.params['acl']
    if module.params['auth']:
        obj_dict['auth'] = module.params['auth']
    if module.params['auth_key']:
        obj_dict['auth_key'] = module.params['auth_key']
    if module.params['key_id']:
        obj_dict['key_id'] = module.params['key_id']

    obj.append(obj_dict)

    return obj


def map_obj_to_commands(want, have, module):

    commands = list()

    server_have = have[0].get('server', None)
    source_int_have = have[0].get('source_int', None)
    acl_have = have[0].get('acl', None)
    logging_have = have[0].get('logging', None)
    auth_have = have[0].get('auth', None)
    auth_key_have = have[0].get('auth_key', None)
    key_id_have = have[0].get('key_id', None)

    for w in want:
        server = w['server']
        source_int = w['source_int']
        [...]
```

However when I run some basic test it I get key error and I ma struggling to find the error

 `File "/tmp/ansible_ios_ntp_payload_80BLjY/__main__.py", line 233, in map_obj_to_commands
KeyError: 'source_int'`

Reply:
I have change the `try` `except` bit with the following. Not sure if that' s exaclty what you were suggesting but I see the code is more clear and like you said it follows the same structure as `have`

```
    server_have = have[0].get('server', None)
    source_int_have = have[0].get('source_int', None)
    acl_have = have[0].get('acl', None)
    logging_have = have[0].get('logging', None)
    auth_have = have[0].get('auth', None)
    auth_key_have = have[0].get('auth_key', None)
    key_id_have = have[0].get('key_id', None)
```